### ISSUE TITLE
adding SubjectAltName to PKCS10 object

### ISSUE THREAD

**Issue by ardigital** [PeculiarVentures/PKI.js#107]
Hi guys - I'm going in circles looking at responses to /issues/90  and /issues/50  - I know i'm close.

Struggling to add SANs to my PKCS10  request the code i have,  where csr = pkijs.simpl.PKCS10()

var san1 = "branch.mydomain.co.uk";
            var san2 = "autodiscover.mydomain.co.uk";

            var extensions = new org.pkijs.simpl.EXTENSIONS();
            var altNames = new org.pkijs.simpl.GENERAL_NAMES({
                names: [
                    new org.pkijs.simpl.GENERAL_NAME({
                        NameType: 1,
                        Name: new org.pkijs.asn1.UNIVERSALSTRING({value: san1})
                    })
                ]
            });

            altNames.names.push(new org.pkijs.simpl.GENERAL_NAME({
                NameType: 1,
                Name: new org.pkijs.asn1.UNIVERSALSTRING({value: san2})
            }));

            extensions.extensions_array.push(new org.pkijs.simpl.EXTENSION({
                extnID: "2.5.29.17", // subjectAltName
                critical: false,
                extnValue: altNames.toSchema().toBER(false),
                parsedValue: altNames
            }));

            var attribute = new org.pkijs.simpl.ATTRIBUTE({
                type: "1.2.840.113549.1.9.14", // pkcs-9-at-extensionRequest
                values: [extensions.toSchema()]
            });

           csr.attributes.push(attribute);

Any feedback much appreciated.
Kind regards
Andy

**Comment by YuryStrozhevsky** [comment]
I should say that PKI.js v1 code is not supported any more. As for additional examples working with SAN you could check issue #94. If it would not help please let me know.

**Comment by ardigital** [comment]
Thanks for getting back so quickly Yury.    Appreciate it's an older version of PKI.js - this is really from learning older examples...  that said,  I look forward to updating the library now I have a better understanding.

Reading notes from #94 - all I'm interpreting is DNS Name is an IA5STRING so when I adjust my code for altNames, as follows:

`var altNames = new org.pkijs.simpl.GENERAL_NAMES({`
	`names: [`
		`new org.pkijs.simpl.GENERAL_NAME({`
			`NameType: 2,	// DNS Name`
			`Name: new org.pkijs.asn1.IA5STRING({value: "autodiscover.mydomain.com" })`
		`})`
	`]`
`});
`
(same applied to the following lines - shortened for example)

When I decode the CSR I'm not seeing an OCTET string for SubjectAltName within the extension 1.2.840.113549.1.9.14

Reading the examples posted by yourself and rmhrisk - you guys both specify AltName as "data.dns" so little puzzled if this implies :

`Name: new org.pkijs.asn1.IA5STRING({value: "autodiscover.mydomain.com" })`


Many thanks

Andy

**Comment by YuryStrozhevsky** [comment]
@ardigital In fact I do not really understand what you want, but I have found a real code from old version of one of commercial appluication using PKI.js v1:
```javascript
        function createCertificateRequest()
        {
            //region Initial variables
            let sequence = Promise.resolve();

            let pkcs10_simpl = new org.pkijs.simpl.PKCS10();

            let hash_algorithm = "sha-1";
            let signature_algorithm_name = "ECDSA";
            //endregion

            //region Get a "crypto" extension
            let crypto = org.pkijs.getCrypto();
            if(typeof crypto == "undefined")
            {
                alert("No WebCrypto extension found");
                return;
            }
            //endregion

            //region Put a static values
            pkcs10_simpl.version = 0;
            pkcs10_simpl.subject.types_and_values.push(new org.pkijs.simpl.ATTR_TYPE_AND_VALUE({ type: "2.5.4.3", value: new org.pkijs.asn1.UTF8STRING({ value: data.name }) }));

            pkcs10_simpl.attributes = new Array();
            //endregion

            //region Create a new key pair
            sequence = sequence.then(
                    function()
                    {
                        //region Get default algorithm parameters for key generation
                        let algorithm = org.pkijs.getAlgorithmParameters(signature_algorithm_name, "generatekey");
                        if("hash" in algorithm.algorithm)
                            algorithm.algorithm.hash.name = hash_algorithm;
                        //endregion

                        return crypto.generateKey(algorithm.algorithm, true, algorithm.usages);
                    }
            );
            //endregion

            //region Store new key in an interim variables
            sequence = sequence.then(
                    function(keyPair)
                    {
                        publicKey = keyPair.publicKey;
                        privateKey = keyPair.privateKey;
                    },
                    function(error)
                    {
                        console.log("Error during key generation: " + error);
                    }
            );
            //endregion

            //region Exporting public key into "subjectPublicKeyInfo" value of PKCS#10
            sequence = sequence.then(
                    function()
                    {
                        return pkcs10_simpl.subjectPublicKeyInfo.importKey(publicKey);
                    }
            );
            //endregion

            //region SubjectKeyIdentifier
            sequence = sequence.then(
                    function(result)
                    {
                        return crypto.digest({ name: "SHA-1" }, pkcs10_simpl.subjectPublicKeyInfo.subjectPublicKey.value_block.value_hex);
                    }
            ).then(
                    function(result)
                    {
                        let extensions = new org.pkijs.simpl.EXTENSIONS({
                            extensions_array: [
                                new org.pkijs.simpl.EXTENSION({
                                    extnID: "2.5.29.14",
                                    critical: false,
                                    extnValue: (new org.pkijs.asn1.OCTETSTRING({ value_hex: result })).toBER(false)
                                })
                            ]
                        });

                        if("email" in data)
                        {
                            let altNames = new org.pkijs.simpl.GENERAL_NAMES({
                                names: [
                                    new org.pkijs.simpl.GENERAL_NAME({
                                        NameType: 1,
                                        Name: data.email
                                    })
                                ]
                            });

                            if("link" in data)
                            {
                                altNames.names.push(new org.pkijs.simpl.GENERAL_NAME({
                                    NameType: 2,
                                    Name: data.link
                                }));
                            }

                            extensions.extensions_array.push(new org.pkijs.simpl.EXTENSION({
                                extnID: "2.5.29.17", // subjectAltName
                                critical: false,
                                extnValue: altNames.toSchema().toBER(false)
                            }));
                        }

                        let attribute = new org.pkijs.simpl.ATTRIBUTE({
                            type: "1.2.840.113549.1.9.14", // pkcs-9-at-extensionRequest
                            values: [extensions.toSchema()]
                        });

                        pkcs10_simpl.attributes.push(attribute);
                    }
            );
            //endregion

            //region Signing final PKCS#10 request
            sequence = sequence.then(
                    function()
                    {
                        return pkcs10_simpl.sign(privateKey, hash_algorithm);
                    },
                    function(error)
                    {
                        console.log("Error during exporting public key: " + error);
                    }
            );
            //endregion

            //region Encode and send final PKCS#10 request
            sequence = sequence.then(
                    function(result)
                    {
                        let requestBuffer = pkcs10_simpl.toSchema(true).toBER(false);

                        let headers = new Headers();
                        headers.append("Content-Type", "application/pkcs10");
                        headers.append("Content-Length", requestBuffer.byteLength);

                        let request = new Request(serverAddress + "/certificate/", {
                            body: requestBuffer,
                            method: "POST",
                            headers: headers
                        });

                        return fetch(request);
                    },
                    function(error)
                    {
                        console.log("Error signing PKCS#10: " + error);
                    }
            );
            //endregion

            //region Parse and check test received user certificate
            sequence = sequence.then(
                    function(result)
                    {
                        return result.arrayBuffer();
                    },
                    function(error)
                    {
                        console.log("Error: " + error);
                    }
            ).then(
                    function(result)
                    {
                        try
                        {
                            let asn1 = org.pkijs.fromBER(result);
                            cert_simpl = new org.pkijs.simpl.CERT({ schema: asn1.result });
                        }
                        catch(ex)
                        {
                            alert("Error getting certificate! Please try again.");
                            return;
                        }
                    }
            );
            //endregion

            return sequence;
        }
```
All "data.<something>" values are simple JS strings. Hope it will help.

**Comment by ardigital** [comment]
Thanks Yury - thoroughly appreciate your assistance.  I had originally tried passing the "data." values as strings, it all looks correct by not showing when I decode the CSR.

I was adapting the example by Charles Engelke https://blog.engelke.com/2015/03/03/creating-x-509-certificates-with-web-crypto-and-pkijs/  

I understand the code snippet above,  I'm going to redo using later version of PKIjs - thanks again for your help.

**Comment by deepzz0** [comment]
@ardigital hi, How do you solve it？
```
var altNames = new AltName({
              altNames: []
            });
            for (var d in domains ) {
              altNames.altNames.push(
                new GeneralName({
                  type: 2,
                  value: new IA5String({ value: d })
              }));
            }
            pkcs10.attributes.push(new Attribute({
                type: '1.2.840.113549.1.9.14', // pkcs-9-at-extensionRequest
                values: [new Extensions({
                    extensions: [new Extension({
                        extnID: '2.5.29.17', // subject alt names
                        critical: false,
                        extnValue: altNames.toSchema().toBER(false)
                    })]
                }).toSchema()]
            }));
```
this example can't work, please tell me how to do this?

**Comment by anindyasundar** [comment]
@YuryStrozhevsky regarding the code you posted on May 30, 2017:
it is running with the V1 example [CSRhelp app](https://github.com/PeculiarVentures/csrhelp).

Now I am trying the same with a V2 example [PKCS#10 complex example](https://pkijs.org/examples/PKCS10_complex_example.html)

```
pkcs10.attributes.push(new Attribute({
				type: "1.2.840.113549.1.9.14", // pkcs-9-at-extensionRequest
				values: [new Extensions({
					extensions: [new Extension({
						extnID: "2.5.29.14",
						critical: false,
						extnValue: new OctetString({ valueHex: result }).toBER(false)
					})]
				}).toSchema()]
			}));
```

I replaced the above code (in region SubjectKeyIdentifier) with the following:

```
var extensions = new Extensions({
                        extensions_array: [new Extension({
                                    extnID: "2.5.29.14",
                                    critical: false,
                                    extnValue: (new OctetString({ valueHex: result })).toBER(false)
                                })]
                        });
						
				var altNames = new GeneralNames({
						 names: [
								  new GeneralName({
										   NameType: 1,
										   Name: "DNS:domain1.com, DNS:domain2.com, DNS:domain3.com"
										   })
								  ]
				});

				 extensions.extensions_array.push(new Extension({
						 extnID: "2.5.29.17", // subjectAltName
						 critical: false,
						 extnValue: altNames.toSchema().toBER(false)
					 }));
				

				 var attribute = new Attribute({
					 type: "1.2.840.113549.1.9.14", // pkcs-9-at-extensionRequest
					 values: [extensions.toSchema()]
				});

				pkcs10.attributes.push(attribute);
```

But this is not running. Can't figure out where I am doing wrong. Please help.

**Comment by YuryStrozhevsky** [comment]
@anindyasundar The "Extensions" class has no "extensions_array" property, only "extensions".

**Comment by anindyasundar** [comment]
@YuryStrozhevsky  Thanks for the reply. I have modified the code as given below. But it still not working.

```
		var extns = new Extensions({
				extensions: [new Extension({
							extnID: "2.5.29.14",
							critical: false,
							extnValue: (new OctetString({ valueHex: result })).toBER(false)
						})]
				});
				
		var altNames = new GeneralNames({
				 names: [
						  new GeneralName({
								   NameType: 1,
								   Name: "DNS:domain1.com, DNS:domain2.com, DNS:domain3.com"
								   })
						  ]
		});

		 extns.extensions.push(new Extension({
				 extnID: "2.5.29.17", // subjectAltName
				 critical: false,
				 extnValue: altNames.toSchema().toBER(false)
			 }));
		

		 var attribute = new Attribute({
			 type: "1.2.840.113549.1.9.14", // pkcs-9-at-extensionRequest
			 values: [extns.toSchema()]
		});

		pkcs10.attributes.push(attribute);		

```
Thank you.

**Comment by YuryStrozhevsky** [comment]
@anindyasundar I made examples for PKIjs and also mentioned examples in the thread of this issue. Please find problems with your code by yourself using these examples and PKIjs source code. 

Thank you.

**Comment by anindyasundar** [comment]
Thanks for the clarification. However, the V1 example code (you posted) adding 'email:' with Subject Alternative Name even if I don't provide an email id. But most common SAN format is "DNS:domain1.com, DNS:domain2.com, DNS:domain3.com".

Is there any way to remove the text 'email:'?

Thank you.

**Comment by YuryStrozhevsky** [comment]
@anindyasundar If you could be more attentive you would find that in the "V1 example" you mentioned there are **TWO** SANs: one for e-mail and one for DNS name. Just use a correct SAN.

Thank you.

**Comment by anindyasundar** [comment]
@YuryStrozhevsky  Got it. Thank you so much! Trying the same with V2.

**Comment by coolaj86** [comment]
@YuryStrozhevsky Thanks for the quick reply.

I'm trying to avoid ES6 and just use plain JavaScript. I'll try to adapt the example back to v1.

**Comment by coolaj86** [comment]
```

**Comment by coolaj86** [comment]
I got this error:

```
Uncaught TypeError: this.extensions_array.toSchema is not a function
    at in_window.org.pkijs.simpl.EXTENSIONS.toSchema (x509_simpl.js:3769)
```

x509_simpl.js:
```
    in_window.org.pkijs.simpl.EXTENSIONS.prototype.toSchema =
    function()
    {
        // #region Construct and return new ASN.1 schema for this object 
        var extension_schemas = new Array();

        for(var i = 0; i < this.extensions_array.length; i++)
            extension_schemas.push(this.extensions_array.toSchema());

        return (new in_window.org.pkijs.asn1.SEQUENCE({
            value: extension_schemas
        }));
        // #endregion 
    }
```

I would think that this might fix it:

```
this.extensions_array.forEach(function (ext) {
  extension_schemas.push(ext.toSchema());
});
```

Does that look right?

**Comment by speedupwebsite** [comment]
@coolaj86 I am posting the working example in 2-3 days.

**Comment by coolaj86** [comment]
@speedupwebsite Thanks.

I found a working version of v1 with the additions that I needed here:

* https://stackoverflow.com/questions/48419456/pki-js-v2-and-and-webcrypto-add-subject-alternative-name-with-csr-certificate/49896425#49896425
* https://getwww.me/V1-csrhelp-master.zip

But it turned out that the v1.2.1 tag, which is the latest on git, is far behind the v1.3 on which the example was based.

My updated version is here: https://coolaj86.com/articles/lets-encrypt-v2-step-by-step/
(mostly I just took stuff out of the other version)

I'll probably publish it as its own module for future reference.
